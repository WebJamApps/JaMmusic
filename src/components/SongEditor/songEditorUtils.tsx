import React from 'react';
import type superagent from 'superagent';
import type { MusicDashboardController } from '../../containers/MusicDashboard/MusicDashboardController';
import type { ISong } from '../../providers/Songs.provider';
import type { MusicDashboard } from '../../containers/MusicDashboard';

async function updateSongAPI(comp:MusicDashboard): Promise<string> {
  const { songState } = comp.state;
  const { editSong } = comp.props;
  const { auth } = comp.props;
  const songChanges = { ...songState, _id: undefined };
  let r: superagent.Response;
  try {
    r = await comp.controller.superagent.put(`${process.env.BackendUrl}/song/${editSong._id}`)
      .set('Content-Type', 'application/json')
      .set('Authorization', `Bearer ${auth.token}`)
      .send(songChanges);
  } catch (e) { return (e as Error).message; }
  if (r.status === 200) { window.location.reload(); return 'song updated'; } // do not reload but instead fetch songs and refresh storage
  return `${r.status} song was not updated`;
}

function resetSongForm(comp:MusicDashboard):void {
  const { dispatch } = comp.props;
  dispatch({
    type: 'EDIT_SONG',
    songData: {
      _id: '', category: '', year: 2021, title: '', url: '',
    },
  });
  comp.setState({
    songState: {
      _id: '', category: '', year: 2021, title: '', url: '',
    },
  });
  window.location.reload();
}

function editSongButtons(comp:MusicDashboard, setNewEditor:any, editSong?: ISong):JSX.Element | null {
  return (editSong && editSong._id !== '' ? (
    <span>
      <button className="floatRight" type="button" id="cancel-edit-song" onClick={() => setNewEditor({ song:{}, tour:{}, image:{} })}>
        Cancel
      </button>
      <button
        className=""
        type="button"
        onClick={() => updateSongAPI(comp)}
      >
        Update
        {' '}
        Song
      </button>
    </span>
  ) : null);
}

// function songForm(controller:MusicDashboardController, songState: ISong, onChangeSong: React.ChangeEventHandler<HTMLInputElement>,
//   handleCategoryChange: (event: React.ChangeEvent<HTMLSelectElement>, isSelected: boolean) => void):JSX.Element {
//   return (
//     <>
//       <label htmlFor="title">
//         * Title
//         <input id="title" value={songState.title} onChange={onChangeSong} />
//       </label>
//       <label htmlFor="url">
//         * Url
//         <input id="url" value={songState.url} onChange={onChangeSong} />
//       </label>
//       <label htmlFor="artist">
//         * Artist
//         <input id="artist" value={songState.artist} onChange={onChangeSong} />
//       </label>
//       <p>* Category</p>
//       {controller.forms.makeDropdown('category', songState.category, handleCategoryChange, ['original', 'mission', 'pub'])}
//     </>
//   );
// }

export const MoreSongForm = (props:{ songState: ISong, onChangeSong: React.ChangeEventHandler<HTMLInputElement> }):JSX.Element => {
  return (
    <>
      <label htmlFor="artist">
        Album
        <input id="album" value={props.songState.album || ''} onChange={props.onChangeSong} />
      </label>
      <label htmlFor="image">
        Image
        <input id="image" value={props.songState.image || ''} onChange={props.onChangeSong} />
      </label>
      <label htmlFor="composer">
        Composer
        <input id="composer" value={props.songState.composer || ''} onChange={props.onChangeSong} />
      </label>
      <label htmlFor="year">
        * Year
        <input type="number" id="year" value={props.songState.year || 2021} onChange={props.onChangeSong} />
      </label>
    </>
  );
};

export const SongForm = (props:{ controller:MusicDashboardController, songState: ISong, onChangeSong: React.ChangeEventHandler<HTMLInputElement>,
  handleCategoryChange: (event: React.ChangeEvent<HTMLSelectElement>, isSelected: boolean) => void }):JSX.Element => {
  console.log(props.songState);
  return (
    <>
      <label htmlFor="title">
        * Title
        <input id="title" value={props.songState.title || ''} onChange={props.onChangeSong} />
      </label>
      <label htmlFor="url">
        * Url
        <input id="url" value={props.songState.url || ''} onChange={props.onChangeSong} />
      </label>
      <label htmlFor="artist">
        * Artist
        <input id="artist" value={props.songState.artist || ''} onChange={props.onChangeSong} />
      </label>
      <p>* Category</p>
      {props.controller.forms.makeDropdown('category', props.songState.category || 'original', 
        props.handleCategoryChange, ['original', 'mission', 'pub'])}
        <MoreSongForm songState={props.songState} onChangeSong={props.onChangeSong}/>
    </>
  );
};

const addSongAPI = async (songState: ISong, auth: { token: string; }, controller: MusicDashboardController): Promise<string> => {
  const newSong = { ...songState, _id: undefined };
  let r: superagent.Response;
  try {
    r = await controller.superagent.post(`${process.env.BackendUrl}/song`)
      .set('Content-Type', 'application/json')
      .set('Authorization', `Bearer ${auth.token}`)
      .send(newSong);
  } catch (e) { return (e as Error).message; }
  if (r.status === 201) { window.location.reload(); return 'song created'; } // do not reload but instead fetch songs and refresh storage
  return `${r.status} song was not created`;
};

const songButtons = (songState: ISong, comp:MusicDashboard, setNewEditor:any, editSong?: ISong): JSX.Element => (
  <div style={{ textAlign: 'left', marginTop: '10px' }}>
    <span style={{
      fontSize: '16px', marginRight: '20px', position: 'relative', display: 'inline-block',
    }}
    >
      <i>* Required</i>
    </span>
    {editSongButtons(comp, setNewEditor, editSong)}
    {!editSong || editSong._id === '' ? (
      <button
        id="add-song-button"
        disabled={!(songState.year && songState.title && songState.url && songState.artist && songState.category)}
        type="button"
        onClick={() => addSongAPI(songState, comp.props.auth, comp.controller)}
      >
        Add Song
      </button>
    ) : null}
  </div>
);

export default {
  updateSongAPI, addSongAPI, resetSongForm, editSongButtons, songButtons,
};
