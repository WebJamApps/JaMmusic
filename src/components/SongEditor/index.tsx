
import React from 'react';
import forms from '../../lib/forms';
import superagent from 'superagent';
import songEditorUtils from './songEditorUtils';
import { EditorContext } from '../../providers/Editor.provider';
import type { ISong } from '../../providers/Songs.provider';

interface Ieditor { song: ISong; tour: Record<string, unknown>; image: Record<string, unknown>; }

export const onChangeSong = (evt: React.ChangeEvent<HTMLInputElement>, editor: Ieditor,
  setNewEditor: (arg0: Ieditor) => void): void => {
  evt.persist();
  const newEditor = { image: {}, tour: {}, song: { ...editor.song, [evt.target.id]: evt.target.value } };
  setNewEditor(newEditor);
};

export const handleCategoryChange = (event: React.ChangeEvent<HTMLSelectElement>, editor: Ieditor,
  setNewEditor: (arg0: Ieditor) => void): void => {
  setNewEditor({ image: {}, tour: {}, song: { ...editor.song, category: event.target.value } });
};

export const makeInput = (required: boolean, id: string,
  editorContext: {
    editor: Ieditor,
    setNewEditor: (arg0: Ieditor) => void
  }): JSX.Element => {
  const { editor, setNewEditor } = editorContext;
  const { song } = editor;
  const songValue: any = song;
  // eslint-disable-next-line security/detect-object-injection
  const inputValue = songValue[id] || '';
  return (<label htmlFor={id}>
    {required ? '* ' : ''}{id}
    <input id={id} value={inputValue} onChange={(evt) => onChangeSong(evt, editor, setNewEditor)} />
  </label>
  );
};

export const MoreSongForm = (props: {
  setNewEditor: (arg0: Ieditor) => void,
  editor: Ieditor,
  onChangeSong: any
}): JSX.Element => {
  return (
    <>
      {makeInput(false, 'album', props)}
      {makeInput(false, 'image', props)}
      {makeInput(false, 'composer', props)}
      {makeInput(true, 'year', props)}
    </>
  );
};

interface IsongFormProps {
  editor: Ieditor,
  setNewEditor: (arg0: Ieditor) => void
}
export const SongForm = (props: IsongFormProps): JSX.Element => {
  const { editor, setNewEditor } = props;
  return (
    <>
      {makeInput(true, 'title', props)}
      {makeInput(true, 'url', props)}
      {makeInput(true, 'artist', props)}
      <p>* Category</p>
      {forms.makeDropdown('category', editor.song.category,
        (evt: React.ChangeEvent<HTMLSelectElement>) => handleCategoryChange(evt,
          editor, setNewEditor), ['original', 'mission', 'pub'])}
      <MoreSongForm setNewEditor={setNewEditor} editor={editor} onChangeSong={onChangeSong} />
    </>
  );
};

export const EditSongButtons = ({ setNewEditor, editor, auth, updateSongAPI }:
{
  setNewEditor: (arg0: Ieditor) => void,
  editor: Ieditor,
  auth: any, updateSongAPI: any
}): JSX.Element => {
  return (
    <span>
      <button className="floatRight" type="button" id="cancel-edit-song"
        onClick={() => setNewEditor({ song: { category: 'original', year: 2022, title: '', url: '' }, tour: {}, image: {} })}>
        Cancel
      </button>
      <button
        className=""
        id="update-song-button"
        type="button"
        onClick={() => updateSongAPI(superagent, editor.song, auth, setNewEditor)}
      >
        Update
        {' '}
        Song
      </button>
    </span>
  );
};

interface IsongButtonsProps {
  setNewEditor: (arg0: Ieditor) => void, editor: Ieditor,
  auth: any, addSongAPI: any, updateSongAPI: any
}
export const SongButtons = ({ editor, setNewEditor, auth, addSongAPI, updateSongAPI }:
IsongButtonsProps): JSX.Element => {
  return (
    <div style={{ textAlign: 'left', marginTop: '10px' }}>
      <span style={{
        fontSize: '16px', marginRight: '20px', position: 'relative', display: 'inline-block',
      }}
      >
        <i>* Required</i>
      </span>
      {editor.song._id ? <EditSongButtons editor={editor} setNewEditor={setNewEditor} auth={auth} updateSongAPI={updateSongAPI} /> :
        <button
          id="add-song-button"
          disabled={!(editor.song.year && editor.song.title && editor.song.url && editor.song.artist)}
          type="button"
          onClick={() => addSongAPI(superagent, editor.song, auth, setNewEditor)}
        >
          Add Song
        </button>
      }
    </div>
  );
};

export const SongFormTitle = ({ editor }: { editor: Ieditor }): JSX.Element => {
  return (
    <h5 style={{ marginBottom: 0 }}>
      {editor.song && editor.song._id && editor.song._id !== '' ? 'Edit ' : 'Add '}
      Song
    </h5>
  );
};

interface IsongEditorDiv {
  editor: Ieditor,
  setNewEditor: (arg0: Ieditor) => void, auth: any
}
export const SongEditorDiv = ({ editor, setNewEditor, auth }: IsongEditorDiv) => {
  if (!editor.song.category) editor.song.category = 'original';
  return (
    <div
      className="material-content elevation3"
      style={{ maxWidth: '320px', margin: '30px auto' }}
    >
      <SongFormTitle editor={editor} />
      <form id="picsForm">
        <SongForm editor={editor} setNewEditor={setNewEditor}
        />
        <p>{' '}</p>
        <SongButtons editor={editor} setNewEditor={setNewEditor} auth={auth} addSongAPI={songEditorUtils.addSongAPI}
          updateSongAPI={songEditorUtils.updateSongAPI} />
      </form>
    </div>
  );
};

export const SongEditor = ({
  auth,
}: { auth: any }): JSX.Element => {
  const { editor, setNewEditor } = React.useContext(EditorContext);
  return <SongEditorDiv editor={editor} setNewEditor={setNewEditor} auth={auth} />;
};
