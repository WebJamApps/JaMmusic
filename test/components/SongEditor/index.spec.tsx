/* eslint-disable prefer-const */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { shallow } from 'enzyme';
import React from 'react';
import SongEditor from '../../../src/components/SongEditor';

describe('SongEditor', () => {
  let editSong:any = {}, songState:any = {},
    comp:any = {}, controller:any = { forms: { makeDropdown: jest.fn() } };

  it('renders correctly in Create mode', () => {
    const wrapper = shallow(<SongEditor
      editSong={editSong}
      songState={songState}
      comp={comp}
      controller={controller}
    />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('h5').text()).toBe('Add Song');
  });
  it('renders in Edit mode', () => {
    editSong = { _id: '123' };
    const wrapper = shallow(<SongEditor
      editSong={editSong}
      songState={songState}
      comp={comp}
      controller={controller}
    />);
    expect(wrapper.find('h5').text()).toBe('Edit Song');
  });
});
