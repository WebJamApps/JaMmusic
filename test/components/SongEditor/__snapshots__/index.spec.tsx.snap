// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SongEditor renders correctly 1`] = `
<div
  className="material-content elevation3"
  style={
    Object {
      "margin": "30px auto",
      "maxWidth": "320px",
    }
  }
>
  <h5
    style={
      Object {
        "marginBottom": 0,
      }
    }
  >
    Add 
    Song
  </h5>
  <form
    id="picsForm"
  >
    <label
      htmlFor="title"
    >
      * 
      title
      <input
        id="title"
        onChange={[Function]}
        value=""
      />
    </label>
    <label
      htmlFor="url"
    >
      * 
      url
      <input
        id="url"
        onChange={[Function]}
        value=""
      />
    </label>
    <label
      htmlFor="artist"
    >
      * 
      artist
      <input
        id="artist"
        onChange={[Function]}
        value=""
      />
    </label>
    <p>
      * Category
    </p>
    <select
      id="category"
      multiple={false}
      onChange={[Function]}
      value="original"
    >
      <option
        id="original"
        value="original"
      >
        original
      </option>
      <option
        id="mission"
        value="mission"
      >
        mission
      </option>
      <option
        id="pub"
        value="pub"
      >
        pub
      </option>
    </select>
    <label
      htmlFor="year"
    >
      * 
      year
      <input
        id="year"
        onChange={[Function]}
        value={2021}
      />
    </label>
    <label
      htmlFor="album"
    >
      
      album
      <input
        id="album"
        onChange={[Function]}
        value=""
      />
    </label>
    <label
      htmlFor="composer"
    >
      
      composer
      <input
        id="composer"
        onChange={[Function]}
        value=""
      />
    </label>
    <label
      htmlFor="image"
    >
      
      image
      <input
        id="image"
        onChange={[Function]}
        value=""
      />
    </label>
    <p>
       
    </p>
    <div
      style={
        Object {
          "marginTop": "10px",
          "textAlign": "left",
        }
      }
    >
      <span
        style={
          Object {
            "display": "inline-block",
            "fontSize": "16px",
            "marginRight": "20px",
            "position": "relative",
          }
        }
      >
        <i>
          * Required
        </i>
      </span>
      <button
        disabled={true}
        id="add-song-button"
        onClick={[Function]}
        type="button"
      >
        Add Song
      </button>
    </div>
  </form>
</div>
`;
